ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,271 @@
+import pickle
+import sqlite3
+from typing import Any, Iterator, MutableMapping, NamedTuple, Tuple, Union
+
+from eth_utils.toolz import first
+
+CREATE_CACHE_QUERY = """
+    CREATE TABLE IF NOT EXISTS cache (
+        id INTEGER NOT NULL PRIMARY KEY,
+        key BLOB NOT NULL UNIQUE,
+        value BLOB NOT NULL,
+        pref INTEGER UNIQUE,
+        nref INTEGER UNIQUE
+    )
+"""
+
+CREATE_MAP_QUERY = """
+    CREATE TABLE IF NOT EXISTS map (
+        key BLOB NOT NULL PRIMARY KEY,
+        pointer INTEGER NOT NULL UNIQUE
+    )
+"""
+
+DEFAULT_CACHE_SIZE = 100000
+
+
+class Node(NamedTuple):
+    id: int
+    key: Union[str, bytes]
+    value: Any
+    pref: int
+    nref: int
+
+
+class LRUSQLDict(MutableMapping[Union[str, bytes], Any]):
+    """
+    SQLite3-backed dictionary that implements an LRU cache.
+    """
+
+    def __init__(
+        self, conn: sqlite3.Connection, cache_size: int = DEFAULT_CACHE_SIZE
+    ) -> None:
+        self._conn = conn
+        self._execute(CREATE_CACHE_QUERY)
+        self._execute(CREATE_MAP_QUERY)
+        self.cache_size = cache_size
+
+    def __iter__(self) -> Iterator[Union[str, bytes]]:
+        with self._conn:
+            for key in self._conn.execute("SELECT key FROM cache"):
+                yield first(key)
+
+    def __len__(self) -> int:
+        result = self._fetch_single_query("SELECT COUNT(*) FROM cache;")
+        return int(first(result))
```

Comment:
I assume the `int` cast is to make `mypy` happy?  If so I think a `# type: ignore` comment would be more appropriate to not incur any runtime overhead (albeit a very small amount).